### DISCUSSION TITLE
Is it possible to use setjmp/longjmp to recover AsyncGetCallTrace from crash

### DISCUSSION THREAD

**Discussion by yanglong1010** [async-profiler/async-profiler#1077] [General]
Hi, 

Currently async-profiler default uses AsyncGetCallTrace. In our environment, our profiler is based on async-profiler 2.8.3, we encountered more than 20 crashes a week related to AsyncGetCallTrace. Some of them are related to known JDK bugs, e.g. [JDK-8283849](https://bugs.openjdk.org/browse/JDK-8283849). There are also other situations though crashed on JDK methods (e.g. forte_fill_call_trace_given_top, frame::sender(RegisterMap*), frame::is_interpreted_frame_valid(JavaThread*)),  but sometimes async-profiler unwind nmethods or stubs, modifies the registers and then recall AsyncGetCallTrace, its hard to reproduce and confirm the bugy parts.

Given that AsyncGetCallTrace replacement [#795](https://github.com/async-profiler/async-profiler/issues/795) is still not marked production ready and we encountered so many hard to reproduce crashes related to AsyncGetCallTrace. We are desperate to find a solution to alleviate the situation, but I have 2 questions blocking me.

I would appreciate it if you can help me ~

- The first is about the --vm cstack solution. I have read the manual pages of [setjmp](https://man7.org/linux/man-pages/man3/setjmp.3.html), [longjmp](https://man7.org/linux/man-pages/man3/longjmp.3.html) and [signal-safety](https://man7.org/linux/man-pages/man7/signal-safety.7.html). Though uses longjmp in signal handlers may corrupt non-async-safe-function related internal structs, but if we detect the signal instruction and only call longjmp at known places, it is safe. Am I right ? Further more, setjmp is not in signal-safety list, is it safe to be called in signal handlers ?
- The second, Inspired by [#795](https://github.com/async-profiler/async-profiler/issues/795), I don't know if we can use setjmp/longjmp to recover AsyncGetCallTrace from segment fault. Async-profiler or java-profiler of Datadog does not use this approach (specifically, Datadog advices the users to enable profiling only on some version and above, e.g. OpenJDK 8u352, 11.0.17+, 17.0.5+, [doc](https://docs.datadoghq.com/profiler/enabling/java/?tab=datadogprofiler)), are there some safety considerations or some factors I didn't consider? e.g. If some code in AsyncGetCallTrace modifies VM state, it is unsafe to longjmp.

Thanks,
Long Yang

**Comment by Jongy** [comment]
The second option was discussed [here](https://github.com/async-profiler/async-profiler/issues/513#issuecomment-1002352496) and @apangin suggested that AGCT is not strictly read-only.

**Comment by apangin** [comment]
Async-profiler 2.8.3 was released 2.5 years ago, there have been _tons_ of fixes and improvements since then.

In general, `cstack=dwarf` helps to workaround certain AsyncGetCallTrace crashes, since it unwinds leaf native frames for the JVM, that otherwise could not do it correctly. You may also turn off additional stack recovery tricks completely by adding `safemode=63` option. This will essentially switch to vanilla AsyncGetCallTrace. However, recent async-profiler version is very careful in stack recovery, it does not do anything known to harm.

If AsyncGetCallTrace could be easily wrapped in crash protection, I would do that long ago. Unfortunately, it isn't. Mainly because it's a black box from async-profiler perspective. It may modify internal JVM state, and it is known to do so. A trivial example: it sets `Thread::_in_asgct` field in the beginning and resets it in the end. If we jump out of the function in the middle of execution, the field won't be reset. There are other examples; see #513 for a previous discussion on the same topic.

Now, to the main point. VMStructs based unwinding is no longer experimental. We are on the path of AsyncGetCallTrace deprecation and making `cstack=vm` the default. See #1073.

**Reply by yanglong1010** [reply]
Thanks for the detailed explanation. Forgive me to repeat a naive question, I see setjmp is not in [signal-safety](https://man7.org/linux/man-pages/man7/signal-safety.7.html) list, is it safe to be called in signal handlers ?

**Reply by apangin** [reply]
Formally, it is not. In practice, any reasonable implementation of setjmp, such as glibc, does nothing more than just saving current execution context (registers) in a provided buffer. As long as no unsafe/non-reentrant code is executed between setjmp/longjmp, or something else that changes global state, there should be no issues.

**Reply by yanglong1010** [reply]
Thanks a lot, very helpful.

**Comment by yanglong1010** [comment]
Refer to the page of [setjmp](https://man7.org/linux/man-pages/man3/longjmp.3.html):
```
setjmp() and longjmp() can be useful for dealing with errors
       inside deeply nested function calls or to allow a signal handler
       to pass control to a specific point in the program, rather than
       returning to the point where the handler interrupted the main
       program.
```
I think it implies `setjmp` is a async-safe-function, but it is not in the signal-safety list, this confuses me a lot.
